src/sage/rings/power_series_ring_element.pyx

```diff
@@ -1135,11 +1136,35 @@ cdef class PowerSeries(AlgebraElement):
             sage: O(x^4)^(1/2)
             O(x^2)
 
+            sage: R.<x,t> = QQ[[]]
+            sage: f = (1+x)^(1+t)
+            sage: f.O(5)
+            1 + x + x*t + 1/2*x^2*t - 1/6*x^3*t + 1/2*x^2*t^2 + O(x, t)^5
+
+            sage: R.<x> = QQ[[]]
+            sage: (e^(e^x - 1)).O(5)
+            1 + x + x^2 + 5/6*x^3 + 5/8*x^4 + O(x^5)
+            sage: e^e^x
+            Traceback (most recent call last):
+            ...
+            ArithmeticError: constant term of power series does not support exponentiation
+
+            sage: R.<x> = QQ[[]]
+            sage: S.<t> = QQ[] # a polynomial ring
+            sage: (1+x)^(1+t)
+            Traceback (most recent call last):
+            ...
+            ValueError: exponent must be a rational number or power series
+
         """
-        try:
-            right = QQ.coerce(r)
-        except TypeError:
-            raise ValueError("exponent must be a rational number")
+        if parent(r) is self.parent() or QQ.has_coerce_map_from(parent(r)):
```

Comment:
I think it would be good to split things up into cases with a bit more care. I would first test if `r` can be coerced into `QQ` and then run that case. Then if that is not the case, check if `r` can be coerced into `self.parent()` and if so then do the $\exp(r \log)$. The reason for the care is that things that coerce into `QQ` might also coerce into `self.parent()` (which I expect the general case to be slower). In both cases, I would probably do it by `try`-`except` blocks.

Reply:
I think is a good solution to manage all the cases